File: src/hetzner_code/commands/destroy.py
from __future__ import annotations

import re
from pathlib import Path

import click

from hetzner_code import github, hetzner, ssh_util, state
from hetzner_code.config import REMOTE_CODE_DIR
from hetzner_code.run import HcodeError


def _parse_git_status(output: str) -> tuple[int, int]:
    """`git status --porcelain=v1 -b` -> (uncommitted_count, ahead_count).
    The first line is always the branch line (e.g. "## main...origin/main
    [ahead 2]"); every line after it is one changed/untracked file."""
    lines = [line for line in output.splitlines() if line]
    if not lines:
        return 0, 0
    match = re.search(r"\[ahead (\d+)", lines[0])
    ahead = int(match.group(1)) if match else 0
    return len(lines) - 1, ahead


def _git_risk(ip: str, label: str, dest: str, identity_path: Path) -> str | None:
    try:
        result = ssh_util.run_remote(
            ip, f"git -C {dest} status --porcelain=v1 -b", identity_path
        )
    except HcodeError:
        return None  # dest doesn't exist, or isn't a git repo — nothing to warn about
    uncommitted, ahead = _parse_git_status(result.stdout)
    parts = []
    if uncommitted:
        parts.append(f"{uncommitted} uncommitted file(s)")
    if ahead:
        parts.append(f"{ahead} unpushed commit(s)")
    return f"{label}: {', '.join(parts)}" if parts else None


def _git_risks(instance: state.Instance) -> list[str]:
    """Everything about to be lost forever once the box is deleted — not
    a full backup, just a heads-up before an irreversible action."""
    if hetzner.describe_server(instance.server_id) is None:
        return []
    identity_path = Path(instance.login_key_path)
    risks = []
    for repo in instance.repos:
        risk = _git_risk(
            instance.ip, repo.name, f"{REMOTE_CODE_DIR}/{repo.name}", identity_path
        )
        if risk:
            risks.append(risk)
        for label in repo.worktrees:
            wt_label = f"{repo.name}-{label}"
            risk = _git_risk(
                instance.ip, wt_label, f"{REMOTE_CODE_DIR}/{wt_label}", identity_path
            )
            if risk:
                risks.append(risk)
    return risks


def _pull_ops_dir(instance: state.Instance) -> None:
    if not (instance.ops_dir and instance.ops_dir_local):
        return
    if hetzner.describe_server(instance.server_id) is None:
        click.echo(f"  (server already gone — couldn't pull {instance.ops_dir} back)")
        return
    click.echo(f"  pulling {instance.ops_dir} -> {instance.ops_dir_local} ...")
    try:
        ssh_util.sync_dir_from(
            instance.ip,
            instance.ops_dir,
            Path(instance.ops_dir_local),
            Path(instance.login_key_path),
        )
    except HcodeError as exc:
        click.echo(f"  warning: couldn't pull ops dir back: {exc}")


def destroy(*, name: str | None, all_: bool, keep_key: bool, yes: bool) -> None:
    if all_:
        targets = state.list_all()
        if not targets:
            click.echo("nothing tracked, nothing to do")
            return
    else:
        if not name:
            raise HcodeError("pass an instance name or --all")
        targets = [state.load(name)]

    for instance in targets:
        repos = ", ".join(r.name for r in instance.repos) or "(no repos)"
        risks = _git_risks(instance)
        prompt = (
            f"Destroy '{instance.name}' ({instance.ip}, {repos})? "
            "This deletes the box and every deploy key on it."
        )
        if risks:
            prompt += "\n  UNSAVED WORK WILL BE LOST:\n" + "\n".join(
                f"    - {r}" for r in risks
            )

        if not yes:
            click.confirm(prompt, abort=True)
        elif risks:
            click.echo(f"warning: destroying '{instance.name}' with unsaved work:")
            for r in risks:
                click.echo(f"  - {r}")

        _pull_ops_dir(instance)

        if not keep_key:
            for repo in instance.repos:
                github.delete_deploy_key(
                    github.RepoRef(owner=repo.owner, name=repo.name), repo.deploy_key_id
                )

        if hetzner.describe_server(instance.server_id) is not None:
            hetzner.delete_server(instance.server_id)
        else:
            click.echo(f"  (server for '{instance.name}' was already gone on Hetzner)")

        state.delete(instance.name)
        click.echo(f"destroyed {instance.name}")

File: internal/commands/destroy.go
package commands

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"

	"github.com/dikee/hcode/internal/config"
	"github.com/dikee/hcode/internal/github"
	"github.com/dikee/hcode/internal/hetzner"
	"github.com/dikee/hcode/internal/run"
	"github.com/dikee/hcode/internal/sshutil"
	"github.com/dikee/hcode/internal/state"
)

var aheadRe = regexp.MustCompile(`\[ahead (\d+)`)

// parseGitStatus turns `git status --porcelain=v1 -b` output into
// (uncommittedCount, aheadCount). The first line is always the branch
// line (e.g. "## main...origin/main [ahead 2]"); every line after it is
// one changed/untracked file.
func parseGitStatus(output string) (int, int) {
	var lines []string
	for _, line := range strings.Split(output, "\n") {
		if line != "" {
			lines = append(lines, line)
		}
	}
	if len(lines) == 0 {
		return 0, 0
	}
	ahead := 0
	if m := aheadRe.FindStringSubmatch(lines[0]); m != nil {
		ahead, _ = strconv.Atoi(m[1])
	}
	return len(lines) - 1, ahead
}

func gitRisk(ip, label, dest, identityPath string) string {
	result, err := sshutil.RunRemote(ip, fmt.Sprintf("git -C %s status --porcelain=v1 -b", dest), identityPath)
	if err != nil {
		return "" // dest doesn't exist, or isn't a git repo — nothing to warn about
	}
	uncommitted, ahead := parseGitStatus(result.Stdout)
	var parts []string
	if uncommitted > 0 {
		parts = append(parts, fmt.Sprintf("%d uncommitted file(s)", uncommitted))
	}
	if ahead > 0 {
		parts = append(parts, fmt.Sprintf("%d unpushed commit(s)", ahead))
	}
	if len(parts) == 0 {
		return ""
	}
	return fmt.Sprintf("%s: %s", label, strings.Join(parts, ", "))
}

// gitRisks is everything about to be lost forever once the box is
// deleted — not a full backup, just a heads-up before an irreversible
// action.
func gitRisks(instance state.Instance) ([]string, error) {
	info, err := hetzner.DescribeServer(instance.ServerID)
	if err != nil {
		return nil, err
	}
	if info == nil {
		return nil, nil
	}
	var risks []string
	for _, repo := range instance.Repos {
		if risk := gitRisk(instance.IP, repo.Name, fmt.Sprintf("%s/%s", config.RemoteCodeDir, repo.Name), instance.LoginKeyPath); risk != "" {
			risks = append(risks, risk)
		}
		for _, label := range repo.Worktrees {
			wtLabel := fmt.Sprintf("%s-%s", repo.Name, label)
			if risk := gitRisk(instance.IP, wtLabel, fmt.Sprintf("%s/%s", config.RemoteCodeDir, wtLabel), instance.LoginKeyPath); risk != "" {
				risks = append(risks, risk)
			}
		}
	}
	return risks, nil
}

func pullOpsDir(instance state.Instance) error {
	if instance.OpsDir == nil || instance.OpsDirLocal == nil {
		return nil
	}
	info, err := hetzner.DescribeServer(instance.ServerID)
	if err != nil {
		return err
	}
	if info == nil {
		fmt.Printf("  (server already gone — couldn't pull %s back)\n", *instance.OpsDir)
		return nil
	}
	fmt.Printf("  pulling %s -> %s ...\n", *instance.OpsDir, *instance.OpsDirLocal)
	if err := sshutil.SyncDirFrom(instance.IP, *instance.OpsDir, *instance.OpsDirLocal, instance.LoginKeyPath); err != nil {
		fmt.Printf("  warning: couldn't pull ops dir back: %s\n", err)
	}
	return nil
}

// DestroyOptions are every flag `hcode destroy` accepts.
type DestroyOptions struct {
	Name    string
	All     bool
	KeepKey bool
	Yes     bool
}

// Destroy runs `hcode destroy`.
func Destroy(opts DestroyOptions) error {
	var targets []state.Instance
	if opts.All {
		all, err := state.ListAll()
		if err != nil {
			return err
		}
		if len(all) == 0 {
			fmt.Println("nothing tracked, nothing to do")
			return nil
		}
		targets = all
	} else {
		if opts.Name == "" {
			return run.Errorf("pass an instance name or --all")
		}
		instance, err := state.Load(opts.Name)
		if err != nil {
			return err
		}
		targets = []state.Instance{instance}
	}

	for _, instance := range targets {
		repoNames := make([]string, len(instance.Repos))
		for i, r := range instance.Repos {
			repoNames[i] = r.Name
		}
		repos := strings.Join(repoNames, ", ")
		if repos == "" {
			repos = "(no repos)"
		}
		risks, err := gitRisks(instance)
		if err != nil {
			return err
		}
		prompt := fmt.Sprintf(
			"Destroy '%s' (%s, %s)? This deletes the box and every deploy key on it.",
			instance.Name, instance.IP, repos,
		)
		if len(risks) > 0 {
			var b strings.Builder
			b.WriteString(prompt)
			b.WriteString("\n  UNSAVED WORK WILL BE LOST:\n")
			for _, r := range risks {
				fmt.Fprintf(&b, "    - %s\n", r)
			}
			prompt = strings.TrimRight(b.String(), "\n")
		}

		if !opts.Yes {
			if err := confirm(prompt); err != nil {
				return err
			}
		} else if len(risks) > 0 {
			fmt.Printf("warning: destroying '%s' with unsaved work:\n", instance.Name)
			for _, r := range risks {
				fmt.Printf("  - %s\n", r)
			}
		}

		if err := pullOpsDir(instance); err != nil {
			return err
		}

		if !opts.KeepKey {
			for _, repo := range instance.Repos {
				if err := github.DeleteDeployKey(github.RepoRef{Owner: repo.Owner, Name: repo.Name}, repo.DeployKeyID); err != nil {
					return err
				}
			}
		}

		info, err := hetzner.DescribeServer(instance.ServerID)
		if err != nil {
			return err
		}
		if info != nil {
			if err := hetzner.DeleteServer(instance.ServerID); err != nil {
				return err
			}
		} else {
			fmt.Printf("  (server for '%s' was already gone on Hetzner)\n", instance.Name)
		}

		if err := state.Delete(instance.Name); err != nil {
			return err
		}
		fmt.Printf("destroyed %s\n", instance.Name)
	}
	return nil
}
